Docs for experiments (#42)
IDinsight/experiments-engine: docs/getting-started/experiments/bayes_ab/index.md

# Bayesian A/B Testing

> Bayesian A/B testing is compares two variants: treatment (e.g. a new feature) and control (e.g. an existing feature). The two variants are shown to users at random. Unlike frequentist A/B testing, this method requires you to set prior probabilities for the treatment and control arms, similarly to the bandit experiments. However, unlike the bandit methods, the posterior is computed at the _end_ of the experiment, and not with every observed outcome. Consequently, this means that the probability with which arms are served to the user are also fixed by the user-defined priors throughout the experiment.

Reviewer: ```suggestion
Bayesian A/B testing compares two variants: treatment (e.g. a new feature) and control (e.g. an existing feature). The two variants are shown to users at random. Unlike frequentist A/B testing, this method lets you set prior probabilities for the treatment and control arms, similarly to the bandit experiments. However, unlike the bandit methods, the posterior is computed at the _end_ of the experiment, and not with every observed outcome. Consequently, this means that the probability with which arms are served to the user are also fixed by the user-defined priors throughout the experiment.
```

> Bayesian A/B testing is compares two variants: treatment (e.g. a new feature) and control (e.g. an existing feature). The two variants are shown to users at random. Unlike frequentist A/B testing, this method requires you to set prior probabilities for the treatment and control arms, similarly to the bandit experiments. However, unlike the bandit methods, the posterior is computed at the _end_ of the experiment, and not with every observed outcome. Consequently, this means that the probability with which arms are served to the user are also fixed by the user-defined priors throughout the experiment.

Author: Hmmm -- I've been thinking about this too. 
There are 2 ways to interpret the prior: (a) the probability of the arm resulting in a desired outcome, OR (b) the probability with which the arm is drawn during the experiment. 
(a) and (b) are the same for bandits, but we can separate the 2 for A/B tests -- I thought we should draw based on the prior, to keep things consistent with the bandits but maybe we can make this a user-configurable decision (i.e. they can choose whether to draw with equal probability or based on the priors).

Will leave as is for now.

## What is Bayesian A/B testing?

> With A/B testing, you have 2 variants of a feature / implementation (one is ideally a baseline / existing feature that you want to compare the other against), with corresponding probability of achieving a desired outcome. You serve beneficiaries either variant at a random but _fixed_ probability throughout the experiment and observe the outcome of their interaction with it. You update these probabilities for each variant _after_ the experiment is over and all te observations are in.

Reviewer: Do we restrict people from adding more than two variants? Can you add additional arms?

Reviewer: This content is very similar to the one above. Can we maybe talk about _why_ or _when_ you'd use this above and get into the details here?

Reviewer: Also, we can peek with bayesian A/B testing right? So you can still see which ARM is doing better. We just don't update sampling probabilities. I'd drop the last line since it's confusing - you have already said that sampling probabilities are fixed during the experiment

> With A/B testing, you have 2 variants of a feature / implementation (one is ideally a baseline / existing feature that you want to compare the other against), with corresponding probability of achieving a desired outcome. You serve beneficiaries either variant at a random but _fixed_ probability throughout the experiment and observe the outcome of their interaction with it. You update these probabilities for each variant _after_ the experiment is over and all te observations are in.

Reviewer: ```suggestion
With A/B testing, you have 2 variants of a feature / implementation (one is ideally a baseline / existing feature that you want to compare the other, a new feature, against). You present users with one of the variants at a random but with a _fixed_ probability throughout the experiment and observe the outcome of their interaction with it. You update these probabilities for each variant _after_ the experiment is over and all the observations are in.
```

## Show me some math!
In our current implementation of Bayesian A/B testing, we use Gaussian priors and support either real-valued or binary outcomes.

The outcome $y$ has a likelihood distribution given by:
$$
\begin{equation}
y \sim p(f(w_{\text{treatment}}\ \mathbb{I}(\text{treatment})\ +\ w_{\text{control}}\ \mathbb{I}(\text{control})\ +\ \text{bias}))
\end{equation}
$$
where $w_{\text{treatment}}$ and $w_{\text{control}}$ denote the treatment and control effect respectively. If $y$ were binary-valued, $p$ denotes the Bernoulli distribution, and $f$ the sigmoid function. For real-valued $y$, $p$ is a Gaussian distribution, $f$ is the identity function.

We also assume Gaussian priors for $w_{\text{treatment}}$ and $w_{\text{control}}$:

$$
\begin{equation}
w_{(\cdot)} \sim \mathcal{N} (\mathbf{\mu}_{(\cdot)}, \Sigma_{(\cdot)})
\end{equation}
$$

> During the course of the experiment, we choose the variant to present to the user with 50\% probability for either arm.

Reviewer: ```suggestion
During the course of the experiment, we choose the variant to present to the user with 50% probability for either arm.
```

> Once we have observed all the outcomes $\[y\]_{j=1}^M$, we can obtain the posterior distribution for the treatment and control arms using the Laplace approximation i.e. the log likelihood can be written down as follows:

Reviewer: ```suggestion
Once we have observed all the outcomes $[y]_{j=1}^M$, we can obtain the posterior distribution for the treatment and control arms using the Laplace approximation i.e. the log likelihood can be written down as follows:
```

$$
\begin{align*}
\log \mathcal{L} &= \sum_{j=1}^M \log p(y_j | f(w_{\text{treatment}}\ \mathbb{I}(\text{treatment})\ +\ w_{\text{control}}\ \mathbb{I}(\text{control})\ +\ \text{bias})) \\
\end{align*}
$$

Then we calculate:

$$
\begin{align*}
\mu_i^{\text(post)} &= \theta_i^* = \underset{\theta_i}{\text{argmax}} \log \mathcal{L}  \\
(\Sigma_i^{\text{post}})^{-1} &= (\Sigma_i^*)^{-1} = \frac{d^2}{d\theta_i^2} \log \mathcal{L} \ \bigg|_{\theta_i^*}
\end{align*}
$$

> We obtain the posterior parameters at the _end_ of the experiment, by using MAPE with the joint likelihood of all the observations.

Reviewer: Annotation to explain MAPE?

> 

Reviewer: Link to "Setting up" and the "Run-experiment" page is missing?
